### PR TITLE
Add left_project() to theseus.geometry.functional.so3

### PR DESCRIPTION
**Pull request by fantaosha**
## Motivation and Context
The `left_project()` operator is original project operator.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: theseus/geometry/functional/lie_group.py [hunk 1]
```diff
@@ -26,6 +26,18 @@ def _jinverse_impl(group: torch.Tensor) -> Tuple[List[torch.Tensor], torch.Tenso
     return _jinverse_impl
 
 
+def LeftProjectImplFactory(module):
+    def _left_project_impl(group: torch.Tensor, matrix: torch.Tensor) -> torch.Tensor:
+        if not module.check_group_tensor(group):
+            raise ValueError("Invalid data tensor for SO3.")
```
Comment (mhmukadam) [inline theseus/geometry/functional/lie_group.py:32]: In the Lie group file should this error message be specific to SO3? This might also be related to a comment in another PR about moving the error message inside the `check_group_tensor` function.

File: theseus/geometry/functional/so3.py [hunk 2]
```diff
@@ -70,6 +70,11 @@ def check_left_act_matrix(matrix: torch.Tensor):
         raise ValueError("Inconsistent shape for the matrix.")
 
 
+def check_left_project_matrix(matrix: torch.Tensor):
+    if matrix.shape[-2:] != (3, 3):
+        raise ValueError("Inconsistent shape for the matrix.")
```
Comment (mhmukadam) [inline theseus/geometry/functional/so3.py:75]: Maybe less ambiguity in the error message about what is 'Inconsistent' would be helpful.

**Review by mhmukadam — COMMENTED** [review]
LGTM!

File: theseus/geometry/functional/so3.py (hunk 2 above)
Comment (luisenp) [reply theseus/geometry/functional/so3.py:75]: Agreed. You can mention what the expected shape is.

File: theseus/geometry/functional/so3.py [hunk 3]
```diff
@@ -702,3 +707,37 @@ def backward(cls, ctx, grad_output):
...
+
+
+# -----------------------------------------------------------------------------
+# Left Project
+# -----------------------------------------------------------------------------
+_left_project_impl = lie_group.LeftProjectImplFactory(_module)
+_jleft_project_impl = None
+
+
+class LeftProject(lie_group.BinaryOperator):
```
Comment (luisenp) [inline theseus/geometry/functional/so3.py:719]: As usual, some comments with a description of what this operator does would be useful. If there is a reference, that's also fine (e.g., "as defined in section XXX of book/paper YYY").

**Review by luisenp — APPROVED** [review]
